tests/slow/src/test/scala/tests/feature/CompletionCrossLspSuite.scala

```diff
@@ -26,7 +26,7 @@ class CompletionCrossLspSuite
   test("serializable-2.13") {
     cleanWorkspace()
     for {
-      _ <- server.initialize(
```

Comment:
Could we instead of adding the hierachy via `initialize` method, just do the class hierachy inside TestServer?
```
trait TestServer { def initialize(...) : Future[Unit] }
class SbtServer extends TestServer { def initialze(...) = ...}
class QuickBuildServer extends TestServer { def initialze(...) = ...}
class BloopServer extends TestServer { def initialze(...) = ...}
```
this would avoid a lot of the changes and reduce the size of the PR.

Reply:
We could also just keep BaseLSPSuite and jut add a constructor parameter on what to use, with default set to quickbuild

Reply:
I don't think that would be any better, because I think it is clearer to keep things separated:
- `TestingServer` is about the Metals language server: what request or command we send and what response we are waiting from it
- `TestingClient` is about the user interaction with the language client: how the user responds to the show message requests
- `workspace` is about writing some files into the filesystem

Whenever we need to coordinate the workspace, the server and the client we should do in the suite level.

If we mix everything inside `TestingServer` then it makes things confusing. (In your proposition `SbtServer` is not really the sbt server, it is the Metals language server in an sbt repository connected to sbt build server. But `BloopServer` could also be in an sbt repository but connected to Bloop and `QuickBuildServer` is also connected to Bloop). So IMO it is simpler to keep things out of `TestingServer`.

So yes that bring a lot of changes but I really think it makes the code easier to navigate and more extensible.

Reply:
> TestingServer is about the Metals language server: what request or command we send and what response we are waiting from it

I don't believe we need to be super strict about it, it can be both language server and build server, why would that be a problem? We can easily add a new server if we need to that. 

> So yes that bring a lot of changes but I really think it makes the code easier to navigate and more extensible.

Is it though? Adding more traits add more things to navigate through and mixing in traits is not really somethings that is easier to reason about. It's much easier to navigate methods than class hierarchy.  Also, I would really keep changes minimal, since it makes reviews quicker. It's a standard in scalafmt reviews and I think it works really good.

But anyway, what do others think about it?

Reply:
I will try to make a full review next week.

Reply:
Okay I did an alternative version https://github.com/scalameta/metals/pull/3006 where `BuildServerInitializer` and `BuildToolLayout` are passed as parameters rather than being mixed in as traits.

Reply:
I mean if the goal is size, then there really isn't much of a difference at all.

Regarding styles, we barely use mixins at all in Metals, which is fine. It typically personally don't like the heavily use them at all. So when I first saw them in here I was skeptical, but I do find the current approach readable and easy-to-navigate. Plus there are places where we maybe avoid class hierarchy too much and end up with some pretty massive files because of it. I don't feel strong on this. This approach is fine to me, but if others strongly prefer the other in #3006, then I have no issue with that either.